test/runtests.jl

```diff
@@ -41,6 +41,22 @@ using UUIDs: uuid4
             @test rows[2].depends_on == rows[1].id
             @test rows[3].depends_on == nothing
             @test rows[4].depends_on == rows[3].id
+
+            # These DAGs hit a method error on Pair{<:TestGenerator, <:TestGenerator}
+            bad_dag = DemoGenerator(1) => DemoGenerator(1)
+            @test_throws TaskFailedException collect(MockTableGenerators.generate(bad_dag))
+
+            bad_dag = DemoGenerator(1) => [DemoGenerator(1) => DemoGenerator(1)]
+            @test_throws TaskFailedException collect(MockTableGenerators.generate(bad_dag))
+
+            # These DAGs hit a method error on Pair{<:TestGenerator, Pair{<:TestGenerator, <:TestGenerator}}
+            # Without this type-restriction the DAGs would be ill-specified and the child
+            # nodes would not inherit the depends_on info from the correct parent.
+            bad_dag = DemoGenerator(1) => DemoGenerator(1) => DemoGenerator(1)
+            @test_throws TaskFailedException collect(MockTableGenerators.generate(bad_dag))
+
+            bad_dag = DemoGenerator(1) => DemoGenerator(1) => [DemoGenerator(1)]
+            @test_throws TaskFailedException collect(MockTableGenerators.generate(bad_dag))
```

Comment:
this is the new stuff. Previously this would return a result but the `depends_on` value for the leaf node would return the grandparent as opposed to their parent.

This might be desireable, but I think this formulation makes it ambiguous which one we want.
(Kind of like trying to evaluate `5 ÷ 3 ÷ 2` without brackets)